DREAM3DReviewFilters/GenerateMaskFromSimpleShapes.cpp

```diff
@@ -0,0 +1,669 @@
+{
+  bool inBounds = false;
+
+  float xmin = xcenter - xwidth / 2.0;
+  float xmax = xcenter + xwidth / 2.0;
+  float ymin = ycenter - yheight / 2.0;
+  float ymax = ycenter + yheight / 2.0;
+  float zmin = zcenter - zdepth / 2.0;
+  float zmax = zcenter + zdepth / 2.0;
+
+  if((x < xmax) && (x > xmin) && (y < ymax) && (y > ymin) && (z < zmax) && (z > zmin))
+  {
+    inBounds = true;
+  }
+
+  return inBounds;
+}
+// -----------------------------------------------------------------------------
+//
+// -----------------------------------------------------------------------------
+bool IsPointInCylinderBounds(float xcenter, float ycenter, float zcenter, float radius, float zdepth, float x, float y, float z)
+{
+  bool inBounds = false;
+
+  float zmin = zcenter - zdepth / 2.0;
+  float zmax = zcenter + zdepth / 2.0;
+
+  if((sqrt((xcenter - x) * (xcenter - x) + (ycenter - y) * (ycenter - y)) < radius) && (z < zmax) && (z > zmin))
+  {
+    inBounds = true;
+  }
+
+  return inBounds;
+}
+// -----------------------------------------------------------------------------
+//
+// -----------------------------------------------------------------------------
+bool IsPointInEllipsoidBounds(float xcenter, float ycenter, float zcenter, float a, float b, float c, float x, float y, float z)
+{
+  bool inBounds = false;
+
+  if(((xcenter - x) * (xcenter - x) / (a * a) + (ycenter - y) * (ycenter - y) / (b * b) + (zcenter - z) * (zcenter - z) / (c * c)) < 1)
+  {
+    inBounds = true;
+  }
+
+  return inBounds;
+}
+// -----------------------------------------------------------------------------
+//
+// -----------------------------------------------------------------------------
+void GenerateMaskFromSimpleShapes::createImageMask()
+{
+  size_t totalNumFIDs = m_CentersPtr.lock()->getNumberOfTuples();
+  size_t totalNumElementsDest = m_MaskPtr.lock()->getNumberOfTuples();
+  DataContainer::Pointer dc = getDataContainerArray()->getDataContainer(getMaskArrayPath().getDataContainerName());
+  ImageGeom::Pointer image = dc->getGeometryAs<ImageGeom>();
+  SizeVec3Type udims = image->getDimensions();
+  FloatVec3Type uorigin = image->getOrigin();
+  FloatVec3Type uspacing = image->getSpacing();
+
+  int64_t dims[3] = {
+      static_cast<int64_t>(udims[0]),
+      static_cast<int64_t>(udims[1]),
+      static_cast<int64_t>(udims[2]),
+  };
+
+  float origin[3] = {
+      static_cast<float>(uorigin[0]),
+      static_cast<float>(uorigin[1]),
+      static_cast<float>(uorigin[2]),
+  };
+
+  float spacing[3] = {
+      static_cast<float>(uspacing[0]),
+      static_cast<float>(uspacing[1]),
+      static_cast<float>(uspacing[2]),
+  };
+
+  bool inBounds = false;
+
+  float currentx = 0;
+  float currenty = 0;
+  float currentz = 0;
+
+  int64_t xindex = 0;
+  int64_t yindex = 0;
+  int64_t zindex = 0;
+  int64_t tempMod = 0;
+
+  size_t k = 0;
+
+  if(m_MaskShape == 0)
```

Comment:
These cases have a lot of duplicated code. They could be simplified with something along the lines of the following:
```
  auto ellipsoidFunc = [this](size_t k, float x, float y, float z) {
    return IsPointInEllipsoidBounds(m_Centers[3 * k], m_Centers[3 * k + 1], m_Centers[3 * k + 2], m_AxisLengths[3 * k], m_AxisLengths[3 * k + 1], m_AxisLengths[3 * k + 2], x, y, z);
  };

... (Insert remaining bounding functions)

  std::function<bool(size_t, float, float, float)> boundingFunc = {};

  switch(m_MaskShape)
  {
  case 0:
    boundingFunc = ellipsoidFunc;
    break;

... (Insert remaining cases)

  }

  for(size_t i = 0; i < totalNumElementsDest; i++)
  {
    int64_t zindex = i / (dims[0] * dims[1]);
    int64_t tempMod = i % (dims[0] * dims[1]);
    int64_t yindex = tempMod / (dims[0]);
    int64_t xindex = tempMod % (dims[0]);

    float currentx = origin[0] + spacing[0] * xindex;
    float currenty = origin[1] + spacing[1] * yindex;
    float currentz = origin[2] + spacing[2] * zindex;

    bool inBounds = false;
    size_t k = 0;

    while(!inBounds && k < totalNumFIDs)
    {
      inBounds = boundingFunc(k, currentx, currenty, currentz);

      if(inBounds)
      {
        m_Mask[i] = true;
      }

      k++;
    }
  }
```

Reply:
If std::function ends up being slow, a templated solution might work better.